### ISSUE TITLE
Design and implement Pipeline Results for pipeline with finally tasks

### ISSUE THREAD

**Issue by pritidesai** [tektoncd/pipeline#2710]
A new field `finally` is being introduced to `PipelineSpec` at the same level as `Tasks`, `Params`, and `Workspaces`. The `finally` field is a list of one or more `PipelineTasks` similar to `Tasks` that provide a guarantee that they are executed only once all `PipelineTasks` under tasks have completed regardless if that is because of success or error.

This new type is getting added with PR #2650 and functionality with PR #2661 

While designing and implementing `finally`, task results were deferred to next iteration. Issue #2557 is tracking task results and in this issue, we will keep track of adding pipeline results to the pipeline with `finally`.

**Comment by bobcatfish** [comment]
@pritidesai has created a WIP design at [Task Results with finally](https://docs.google.com/document/d/10iEJqVstY6k3KNvAXgffIJLcHRbPQ-GIAfQk5Dlrf3c/edit#)

**Comment by pritidesai** [comment]
Reopening as this feature is not implemented yet. And will be a good feature to have finally task produce some kind of result and have it as part of the pipeline.

**Comment by jerop** [comment]
/lifecycle frozen

may be useful even for pipelines in pipelines for the results from finally tasks to be passed to pipeline results

**Comment by jerop** [comment]
Is the idea here that a `Pipeline`'s `Result` can reference a `Finally Task`'s `Result` using a variable `$(finally.<pipeline-task-name>.results.<result-name>)`?

Was curious to see what happens if a a `Pipeline`'s `Result` references a `Finally Task`'s `Result` using a variable `$(tasks.<pipeline-task-name>.results.<result-name>)` - and found that it worked, is that expected behavior? If so, can we document this and close this issue? If not, can we fix it and provide the finally results in pipeline level, or do users depend on the existing behavior? 

#### Example

Modified the `Pipeline` in the provided [example][example] as such:
```yaml
# Pipeline to clone repo into shared workspace and cleanup the workspace after done
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: clone-cleanup-workspace
spec:
  results:
    - name: initialized
      value: $(tasks.check-git-commit.results.init)
  workspaces:
    # common workspace where git repo is cloned and needs to be cleanup after done
    - name: git-source
  tasks:
    # Clone app repo to workspace
    - name: clone-app-repo
      taskRef:
        name: git-clone-from-catalog
      params:
        - name: url
          value: https://github.com/tektoncd/community.git
        - name: subdirectory
          value: application
      workspaces:
        - name: output
          workspace: git-source
  finally:
    # Cleanup workspace
    - name: cleanup
      taskRef:
        name: cleanup-workspace
      workspaces:
        - name: source
          workspace: git-source
    - name: check-git-commit
      params:
        - name: commit
          value: $(tasks.clone-app-repo.results.commit)
      taskSpec:
        params:
          - name: commit
        results:
          - name: init
        steps:
          - name: check-commit-initialized
            image: alpine
            script: |
              if [[ ! $(params.commit) ]]; then
                exit 1
              fi
              printf true | tee /tekton/results/init
```

And this was the created `PipelineRun`
```shell
$ tkn pr describe -L
Name:              write-and-cleanup-workspace
Namespace:         default
Pipeline Ref:      clone-cleanup-workspace
Service Account:   default
Timeout:           1h0m0s
Labels:
 tekton.dev/pipeline=clone-cleanup-workspace

🌡️  Status

STARTED          DURATION     STATUS
42 seconds ago   26 seconds   Succeeded

📦 Resources

 No resources

⚓ Params

 No params

📝 Results

 NAME            VALUE
 ∙ initialized   true

📂 Workspaces

 NAME           SUB PATH   WORKSPACE BINDING
 ∙ git-source   ---        VolumeClaimTemplate

🗂  Taskruns

 NAME                                             TASK NAME          STARTED          DURATION     STATUS
 ∙ write-and-cleanup-workspace-check-git-commit   check-git-commit   24 seconds ago   7 seconds    Succeeded
 ∙ write-and-cleanup-workspace-cleanup            cleanup            24 seconds ago   8 seconds    Succeeded
 ∙ write-and-cleanup-workspace-clone-app-repo     clone-app-repo     42 seconds ago   18 seconds   Succeeded

⏭️  Skipped Tasks

 No Skipped Tasks
```

[example]: https://github.com/tektoncd/pipeline/blob/9605e3e7f31b02d44220c15592b7a0a211f8a0fb/examples/v1beta1/pipelineruns/pipelinerun-with-final-tasks.yaml

**Comment by jerop** [comment]
This was fixed by @vsinghai in https://github.com/tektoncd/community/blob/main/teps/0116-referencing-finally-task-results-in-pipeline-results.md - thank you!
